src/main/java/org/thoughtcrime/securesms/service/FetchForegroundService.java

```diff
@@ -30,15 +32,40 @@ public static void start(Context context) {
     }
 
     GenericForegroundService.createFgNotificationChannel(context);
-    synchronized (SERVICE_LOCK) {
-      if (service == null) {
-        service = new Intent(context, FetchForegroundService.class);
-        ContextCompat.startForegroundService(context, service);
+    try {
+      synchronized (SERVICE_LOCK) {
+        if (service == null) {
+          service = new Intent(context, FetchForegroundService.class);
+          ContextCompat.startForegroundService(context, service);
+        }
+      }
+    } catch (Exception e) {
+      Log.w(TAG, "Failed to start foreground service: " + e + ", fetching in background.");
+      // According to the documentation https://firebase.google.com/docs/cloud-messaging/android/receive,
+      // we need to handle the message within 20s, and the time window may be even shorter than 20s,
+      // so, use 10s to be safe.
+      fetchingSynchronously = true;
+      if (ApplicationContext.dcAccounts.backgroundFetch(10)) {
+        // The background fetch was successful, but we need to wait until all events were processed.
+        // After all events were processed, we will get DC_EVENT_ACCOUNTS_BACKGROUND_FETCH_DONE,
+        // and stop() will be called.
+        while (fetchingSynchronously) {
+          try {
+            // The `wait()` needs to be enclosed in a while loop because there may be
+            // "spurious wake-ups", i.e. `wait()` may return even though `notifyAll()` wasn't called.
+            STOP_NOTIFIER.wait();
+          } catch (InterruptedException ex) { }
+        }
```

Comment:
In order to simplify the logic here, we could also simply call Thread.sleep(1000) and assume that the UI will be done with showing a notification within 1s.

Reply:
I don't think the logic is too complicated, and it is well documented, better than some fixed sleeping that might be too much in some cases and too little in another